src/tools/perf/lib/libperf.c

```diff
@@ -1012,6 +1030,64 @@ ucp_perf_test_rkey_unpack(ucx_perf_thread_context_t *thread,
     return UCS_OK;
 }
 
+static ucs_status_t ucp_perf_test_setup_self_endpoint(ucx_perf_context_t *perf)
+{
+    void *rkey_buffer   = NULL;
+    size_t rkey_size    = 0;
+    ucp_worker_h worker = perf->ucp.worker;
+    size_t address_length;
+    ucp_address_t *address;
+    ucs_status_t status;
+    ucp_ep_params_t ep_params;
+
+    status = ucp_worker_get_address(worker, &address, &address_length);
+    if (status != UCS_OK) {
+        ucs_error("ucp_worker_get_address() failed: %s",
+                  ucs_status_string(status));
+        return status;
+    }
+    ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+    ep_params.address    = address;
+
+    if (perf->params.flags & UCX_PERF_TEST_FLAG_ERR_HANDLING) {
+        ep_params.field_mask     |= UCP_EP_PARAM_FIELD_ERR_HANDLER |
+                                    UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+        ep_params.err_handler.cb  = ucp_perf_test_err_handler;
+        ep_params.err_handler.arg = NULL;
+        ep_params.err_mode        = UCP_ERR_HANDLING_MODE_PEER;
+    }
+
+    status = ucp_ep_create(worker, &ep_params, &perf->ucp.self_ep);
+
+    if (status != UCS_OK) {
+        if (perf->params.flags & UCX_PERF_TEST_FLAG_VERBOSE) {
+            ucs_error("ucp_ep_create() failed: %s", ucs_status_string(status));
+        }
+        return status;
+    }
+    status = ucp_rkey_pack(perf->ucp.context, perf->ucp.recv_memh, &rkey_buffer,
```

Comment:
1. space line before, not after
2. why self_rkey is for recv_memh - what happens if we need to copy to send buffer?